clang/include/clang/Sema/Sema.h

```diff
@@ -12012,6 +12012,30 @@ class Sema final {
     ConstructorDestructor,
     BuiltinFunction
   };
+
+  /// Creates a DeviceDiagBuilder that emits the diagnostic if the current
+  /// context is "used as device code".
+  ///
+  /// - If CurLexicalContext is a kernel function or it is known that the
+  ///   function will be emitted for the device, emits the diagnostics
+  ///   immediately.
+  /// - If CurLexicalContext is a function and we are compiling
+  ///   for the device, but we don't know that this function will be codegen'ed
+  ///   for devive yet, creates a diagnostic which is emitted if and when we
+  ///   realize that the function will be codegen'ed.
+  ///
+  /// Example usage:
+  ///
+  /// Variables with thread storage duration are not allowed to be used in SYCL
+  /// device code
+  /// if (getLangOpts().SYCLIsDevice)
+  ///   SYCLDiagIfDeviceCode(Loc, diag::err_thread_unsupported);
+  DeviceDiagBuilder SYCLDiagIfDeviceCode(SourceLocation Loc, unsigned DiagID);
+
+  /// Checks if Callee function is a device function and emits
```

Comment:
This description isn't particularly clear to me.  Is this for the call-graph based diagnostics here in SemaSYCL (rather than normal SEMA)?  If so, I would think this should be going away as soon as we can make it so.

Reply:
I actually saw this one earlier and have a hard time understanding the purpose of it.

Reply:
Yeah, the description is a bit nasty, maybe I will rewrite it a bit...
Actually, this function is not about call-graph based diagnostics, it's for deferred diagnostics. Deferred diagnostics system also builds some callgraph, but it's not real `CallGraph` object, it's some `DenseMap` which helps to understand where device code sits.
CUDA and OpenMP also have functions like this (as far as I remember for CUDA it has name `CheckCUDACall`, but note that such functions has additional calls diagnosing which is not needed for SYCL).